tokio/src/syscall/udp.rs

```diff
@@ -0,0 +1,336 @@
+use std::{
+    convert::TryFrom,
+    io, net,
+    task::{Context, Poll},
+};
+
+cfg_syscall! {
+    use crate::syscall::Syscalls;
+    use std::sync::Arc;
+    /// A representation of a UDP socket.
+    #[derive(Debug, Copy, Clone)]
+    pub struct UdpResource(usize);
+
+    fn syscalls() -> Arc<dyn Syscalls> {
+        crate::runtime::context::syscalls().expect("Syscalls not supplied to the Runtime")
+    }
+}
+cfg_not_syscall! {
+    use crate::io::PollEvented;
+    /// A representation of a UDP socket.
+    pub type UdpResource = PollEvented<mio::net::UdpSocket>;
+}
+
+impl UdpResource {
+    pub(crate) fn bind_addr(addr: &net::SocketAddr) -> io::Result<Self> {
+        cfg_if_syscall! {
+            {
+                syscalls().udp_bind_addr(addr)
+            } else {
+                let sys = mio::net::UdpSocket::bind(&addr)?;
+                let io = PollEvented::new(sys)?;
+                Ok(io)
+            }
+        }
```

Comment:
If I understand the `cfg_if_syscall!` macro correctly, I _think_ the usages in this module could be
```suggestion
        cfg_if_syscall! {
            syscalls().udp_bind_addr(addr)
        } else {
            let sys = mio::net::UdpSocket::bind(&addr)?;
            let io = PollEvented::new(sys)?;
            Ok(io)
        }
```
which is a bit less awkward...

Reply:
Hmm currently the `cfg_if_syscall` expects two blocks, one before and one after the `else`. I'm not sure how we could get rid of the first block.